src/t8_schemes/t8_default/t8_dpyramid_bits.c

```diff
@@ -0,0 +1,1498 @@
+  default:
+    SC_ABORT_NOT_REACHED ();
+  }
+}
+
+/*Copies a pyramid from source to dest*/
+void
+t8_dpyramid_copy (const t8_dpyramid_t * source, t8_dpyramid_t * dest)
+{
+  if (source == dest) {
+    return;
+  }
+  memcpy (dest, source, sizeof (t8_dpyramid_t));
+}
+
+int
+t8_dpyramid_compare (const t8_dpyramid_t * p1, const t8_dpyramid_t * p2)
+{
+  int                 maxlvl;
+  t8_linearidx_t      id1, id2;
+  T8_ASSERT (p1->x >= 0 && p1->y >= 0 && p1->z >= 0 &&
+             p1->level >= 0 && p1->level <= T8_DPYRAMID_MAXLEVEL);
+  T8_ASSERT (p2->x >= 0 && p2->y >= 0 && p2->z >= 0 &&
+             p2->level >= 0 && p2->level <= T8_DPYRAMID_MAXLEVEL);
+  maxlvl = SC_MAX (p1->level, p2->level);
+
+  id1 = t8_dpyramid_linear_id (p1, maxlvl);
+  id2 = t8_dpyramid_linear_id (p2, maxlvl);
+  if (id1 == id2) {
+    /* The linear ids are the same, the pyramid with the smaller level
+     * is considered smaller */
+    return p1->level - p2->level;
+  }
+  /* return negative if id1 < id2, zero if id1 = id2, positive if id1 >
+     id2 */
+  return id1 < id2 ? -1 : id1 != id2;
+}
+
+int
+t8_dpyramid_get_level (const t8_dpyramid_t * p)
+{
+  T8_ASSERT (0 <= p->level && p->level <= T8_DPYRAMID_MAXLEVEL);
+  return p->level;
+}
+
+int
+t8_dpyramid_custom_mod (t8_linearidx_t * id, const t8_dpyramid_type_t type,
+                        const t8_linearidx_t pyra, const t8_linearidx_t tet)
+{
+  t8_linearidx_t      test = 0, shift;
+  T8_ASSERT (id != NULL);
+  T8_ASSERT (*id >= 0);
+  int                 remain = -1;
+  do {
+    /* Iterate through the local-id. Get the current shift by the type of the
+     * current element*/
+    shift =
+      t8_dpyramid_parenttype_Iloc_to_type[type][remain + 1] >= 6 ? pyra : tet;
+    /*Add the shift to test */
+    test += shift;
+    remain++;
+  } while (test <= (*id));
+  /*test is now larger than id, subtract last shift from test */
+  test -= shift;
+  /*Compute the remaining ID */
+  (*id) -= test;
+  T8_ASSERT (0 <= remain && remain < T8_DPYRAMID_CHILDREN);
+  return remain;
+}
+
+void
+t8_dpyramid_init_linear_id (t8_dpyramid_t * p, const int level,
+                            t8_linearidx_t id)
+{
+  t8_dpyramid_type_t  type;
+  t8_linearidx_t      local_index,
+    p_sum1 = ((t8_linearidx_t) 1) << (3 * level),
+    p_sum2 = sc_intpow64u (6, level);
+  t8_dpyramid_cube_id_t cube_id;
+  int                 i, offset_expo;
+
+  T8_ASSERT (0 <= level && level <= T8_DPYRAMID_MAXLEVEL);
+  T8_ASSERT (0 <= id && id <= 2 * p_sum1 - p_sum2);
+
+  p->level = level;
+  p->x = 0;
+  p->y = 0;
+  p->z = 0;
+  type = 6;                     /*This is the type of the root pyramid */
+  for (i = 1; i <= level; i++) {
+    offset_expo = T8_DPYRAMID_MAXLEVEL - i;
+    p_sum1 >>= 3;
+    p_sum2 /= 6;
+    // Thy types of the tetrahedron children of pyramid are always 0 or 3
+    if (type == 0 || type == 3) {
+      t8_dtet_init_linear_id_with_level ((t8_dtet_t *) p, id, i, level, type);
```

Comment:
When calling tet functions with a pyramid we should not cast explitcitely to a t8_dtet_t:
```suggestion
      t8_dtet_init_linear_id_with_level (p, id, i, level, type);
```

Why? The whole logic relies on a tet and a pyramid being the same struct and the function call only works in that case.
If we have the explicit cast then we will not get warned in the case when the pyramid an tet are not the same struct anymore.
This may occur when some future developer decides to change the implementation of a pyramid.
In this case, the code should definitely not compile, since then the tet logic breaks.
If we use casts however, the code will compile and produce wrong results.